diff --git a/railties/lib/commands/server.rb b/railties/lib/commands/server.rb
index ebe34a4..1199d22 100644
--- a/railties/lib/commands/server.rb
+++ b/railties/lib/commands/server.rb
@@ -55,12 +55,12 @@ puts "=> Booting #{ActiveSupport::Inflector.demodulize(server)}"
 puts "=> Rails #{Rails.version} application starting on http://#{options[:Host]}:#{options[:Port]}#{options[:path]}"
 
 %w(cache pids sessions sockets).each do |dir_to_make|
-  FileUtils.mkdir_p(File.join(RAILS_ROOT, 'tmp', dir_to_make))
+  FileUtils.mkdir_p(File.join(Rails.configuration.tmp_dir, dir_to_make))
 end
 
 if options[:detach]
   Process.daemon
-  pid = "#{RAILS_ROOT}/tmp/pids/server.pid"
+  pid = File.join(Rails.configuration.tmp_dir, 'pids', 'server.pid')
   File.open(pid, 'w'){ |f| f.write(Process.pid) }
   at_exit { File.delete(pid) if File.exist?(pid) }
 end
diff --git a/railties/lib/initializer.rb b/railties/lib/initializer.rb
index a04405a..66be36f 100644
--- a/railties/lib/initializer.rb
+++ b/railties/lib/initializer.rb
@@ -694,6 +694,10 @@ Run `rake gems:install` to install the missing gems.
     # (e.g. log/development.log or log/production.log).
     attr_accessor :log_path
 
+    # The path to use for temporary files. Defaults to tmp/. Cache
+    # files, sockets, session files, and pid files are kept here.
+    attr_accessor :tmp_dir
+
     # The specific logger to use. By default, a logger will be created and
     # initialized using #log_path and #log_level, but a programmer may
     # specifically set the logger to use via this accessor and it will be
@@ -819,6 +823,7 @@ Run `rake gems:install` to install the missing gems.
       self.eager_load_paths             = default_eager_load_paths
       self.log_path                     = default_log_path
       self.log_level                    = default_log_level
+      self.tmp_dir                      = default_tmp_dir
       self.view_path                    = default_view_path
       self.controller_paths             = default_controller_paths
       self.preload_frameworks           = default_preload_frameworks
@@ -991,6 +996,10 @@ Run `rake gems:install` to install the missing gems.
         environment == 'production' ? :info : :debug
       end
 
+      def default_tmp_dir
+        File.join(root_path, 'tmp')
+      end
+
       def default_database_configuration_file
         File.join(root_path, 'config', 'database.yml')
       end
@@ -1044,8 +1053,8 @@ Run `rake gems:install` to install the missing gems.
       end
 
       def default_cache_store
-        if File.exist?("#{root_path}/tmp/cache/")
-          [ :file_store, "#{root_path}/tmp/cache/" ]
+        if File.exist? File.join(default_tmp_dir, 'cache')
+          [ :file_store, File.join(default_tmp_dir, 'cache') ]
         else
           :memory_store
         end